plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/UriResourceMap.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2019 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ide.xtext.server;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.Resource;
+
+import com.google.common.collect.Collections2;
+import com.google.common.collect.ForwardingMap;
+import com.google.inject.Inject;
+
+/**
+ * A project specific map of resources per URI. It acts as a mere filter to the {@link GlobalUriResourceMap} such that
+ * an invocation of {@link #clear()} does not wipe out the entire map contents but only the resources from the current
+ * project.
+ */
+public class UriResourceMap extends ForwardingMap<URI, Resource> {
```

Comment:
Could the class name reflect that this map is reading and writing to a global state, but deleting only those entries which it previously added?

Reply:
Hmm ... it probably should. What do you think about `ProjectAwareUriResourceMap`?

Reply:
Seems to me "Adapter" should be in the name, because it does not contain any key/value pairs itself and acts only as an adapter of the `GlobalUriResourceMap`.

What about `ProjectSpecificGlobalUriResourceMapAdapter` (I know, it's long ;-) )

Reply:
Offline discussion: Use `WorkspaceUriResourceMap` and `ProjectUriResourceMap`